### PR TITLE
include spaces etags in virtual spaces

### PR DESCRIPTION
**Pull request by C0rby**
The shares storage provider didn't include the etag in virtual spaces like the shares jail or mountpoints.

### CODE REVIEW

File: internal/grpc/services/gateway/usershareprovider.go [hunk 1]
```diff
@@ -160,6 +162,48 @@ func (s *svc) ListReceivedShares(ctx context.Context, req *collaboration.ListRec
...
+		shareEtags[rs.Share.Id.OpaqueId] = sRes.Info.Etag
+	}
+
+	marshalled, err := json.Marshal(shareEtags)
+	if err != nil {
+		logger.Error().
+			Err(err).
+			Msg("ListRecievedShares: failed marshal share etags")
+	} else {
+		opaque := res.Opaque
```
Comment (micbar) [inline internal/grpc/services/gateway/usershareprovider.go:194]: You could use `utils.AddPlainToQpaque()`

Comment (micbar) [inline internal/grpc/services/gateway/usershareprovider.go:194]: You could use `utils.AddPlainToQpaque()`

File: internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go [hunk 2]
```diff
@@ -380,8 +390,36 @@ func (s *service) ListStorageSpaces(ctx context.Context, req *provider.ListStora
...
+					case current.Mtime.Seconds == earliestShare.Mtime.Seconds &&
+						current.Mtime.Nanos > earliestShare.Mtime.Nanos:
+						earliestShare = current
+					}
+				}
 
+				var opaque *typesv1beta1.Opaque
+				if earliestShare != nil {
+					if etag, ok := shareEtags[earliestShare.Id.OpaqueId]; ok {
+						opaque = &typesv1beta1.Opaque{
```
Comment (micbar) [inline internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go:411]: you could use `utils.AddPlainToOpaque()`

Comment (micbar) [inline internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go:411]: you could use `utils.AddPlainToOpaque()`

File: internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go [hunk 3]
```diff
@@ -401,8 +439,20 @@ func (s *service) ListStorageSpaces(ctx context.Context, req *provider.ListStora
 					// none of our business
 					continue
 				}
+				var opaque *typesv1beta1.Opaque
+				if etag, ok := shareEtags[receivedShare.Share.Id.OpaqueId]; ok {
+					opaque = &typesv1beta1.Opaque{
```
Comment (micbar) [inline internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go:444]: same

Comment (micbar) [inline internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go:444]: same

File: internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go [hunk 4]
```diff
@@ -437,7 +487,19 @@ func (s *service) ListStorageSpaces(ctx context.Context, req *provider.ListStora
 						continue
 					}
 				}
+				var opaque *typesv1beta1.Opaque
+				if etag, ok := shareEtags[receivedShare.Share.Id.OpaqueId]; ok {
```
Comment (micbar) [inline internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go:491]: same

Comment (micbar) [inline internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go:491]: same

**Review by micbar — COMMENTED** [review]
See comments

File: internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go [hunk 5]
```diff
@@ -360,8 +359,9 @@ func (s *service) ListStorageSpaces(ctx context.Context, req *provider.ListStora
 	}
 
 	var receivedShares []*collaboration.ReceivedShare
+	var shareEtags map[string]string
 	if fetchShares {
-		lsRes, err := s.sharesProviderClient.ListReceivedShares(ctx, &collaboration.ListReceivedSharesRequest{
+		lsRes, err := s.gateway.ListReceivedShares(ctx, &collaboration.ListReceivedSharesRequest{
```
Comment (kobergj) [inline internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go:364]: This change will cause a worse performance. Before it we had only one call from sharesstorageprovider to `sharesProvider`. Now it will call gateway inbetween which will then call `sharesProvider`. I understand why this change is necessary but I am a little concerned about performance. 

Should we cache `ListReceivedShares` calls on gateway?

File: internal/grpc/services/gateway/usershareprovider.go [hunk 6]
```diff
@@ -160,6 +162,48 @@ func (s *svc) ListReceivedShares(ctx context.Context, req *collaboration.ListRec
 	if err != nil {
 		return nil, errors.Wrap(err, "gateway: error calling ListReceivedShares")
 	}
+
+	// TODO: This is a hack for now.
+	// Can we do that cleaner somehow?
```
Comment (kobergj) [inline internal/grpc/services/gateway/usershareprovider.go:167]: Potentially we could use events to propagate e-tags also to share roots. But I'm not entirely sure that's feasible.

File: internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go (hunk 5 above)
Comment (C0rby) [reply internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go:364]: > I understand why this change is necessary but I am a little concerned about performance.

Yeah, I understand but let's implement the stuff first now and care about performance optimization later (in the beta phase).

File: internal/grpc/services/gateway/usershareprovider.go (hunk 6 above)
Comment (C0rby) [reply internal/grpc/services/gateway/usershareprovider.go:167]: That'd be possible, but required some CS3 API changes to somehow include the etag in a share.

File: internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go (hunk 5 above)
Comment (kobergj) [reply internal/grpc/services/sharesstorageprovider/sharesstorageprovider.go:364]: Yes sounds good :+1:

File: internal/grpc/services/gateway/usershareprovider.go (hunk 6 above)
Comment (kobergj) [reply internal/grpc/services/gateway/usershareprovider.go:167]: Yes. I was only about possible future improvements. The hack is good for now :+1:
